### PR TITLE
bthrm - Move caching of characteristics into settings

### PR DESCRIPTION
**Pull request by halemmerich**
Moves the caching of device characteristics out of the library code and into the settings. Reduces the lib by about 1/6 and potentially saves on flash writes since the connect is now strictly read only. Caching now only happens once during initial pairing with the sensor.

### CODE REVIEW

**Comment by thyttan** [comment]
Under the files tab there are new lint warnings: `Unchanged files with check annotations `

```
/home/runner/work/BangleApps/BangleApps/apps/bthrm/lib.js
Warning:   331:9  warning  'attachServicePromise' is assigned a value but never used  no-unused-vars

/home/runner/work/BangleApps/BangleApps/apps/bthrm/settings.js
Warning:   24:9  warning  'bthrm' is assigned a value but never used  no-unused-vars
```

Do you think it makes sense to add a commit fixing those on this PR?

**Comment by halemmerich** [comment]
Yes I think both of those are just forgotten. `attachServicePromise` probably makes other stuff obsolete on removal as well. Will do that.

File: apps/bthrm/lib.js [hunk 1]
```diff
@@ -316,31 +280,32 @@ exports.enable = () => {
...
-    var attachCharacteristicPromise = function(promise, characteristic) {
+    let attachCharacteristicPromise = function(promise, characteristic) {
       return promise.then(()=>{
         log("Handling characteristic:", characteristic);
         return createCharacteristicPromise(characteristic);
       });
     };
 
-    var createCharacteristicsPromise = function(newCharacteristics) {
+    let createCharacteristicsPromise = function(newCharacteristics) {
```
Comment (thyttan) [inline apps/bthrm/lib.js:305]: Like you suspected - this function is no longer used as well now.

**Comment by thyttan** [comment]
- I feel like I see very big improvements in time to connection and getting readings when in the BT HRM app. It's even quicker than the internal hrm! (v0.18 seems to not get a connection for me at all on fw2v21.82)

- The `recorder` works with default mode of `bthrm`. But I don't get readings for the `recorder` "Log BT HR" entry when using "both" mode of `bthrm`.

- If I run the `bthrm` app on the watch and then connect with the Web IDE I get errors and the app hangs:

```
        var e=null==d.bgCol?a:g.toColor(d.bgCol);if(e!=a||"txt"==d.t...
        ^
debug>
In debug mode: Expected a simple ID, type 'help' for more info.
Found BANGLEJS2, 2v21.82
ERROR: Ctrl-C while processing interval - removing it.
Execution Interrupted during event processing.
Interpreter error: [
  "CALLBACK"
 ]
New interpreter error: CALLBACK
>
Connected to Web Bluetooth, Bangle.js 92f9
>
{"t":"status","bat":16,"chg":0}
{"t":"ver","fw":"2v21.82","hw":2}
{"t":"force_calendar_sync","ids":[]}
Disconnected from Web Bluetooth, Bangle.js 92f9
        var e=null==d.bgCol?a:g.toColor(d.bgCol);if(e!=a||"txt"==d.t...
        ^
debug>
In debug mode: Expected a simple ID, type 'help' for more info.
ERROR: Ctrl-C while processing interval - removing it.
Execution Interrupted during event processing.
Interpreter error: [
  "CALLBACK"
 ]
New interpreter error: CALLBACK
```

**Comment by halemmerich** [comment]
> * I feel like I see very big improvements in time to connection and getting readings when in the BT HRM app. It's even quicker than the internal hrm! (v0.18 seems to not get a connection for me at all on fw2v21.82)

Great, that was kind of the reason for refactoring :)

> 
>     * The `recorder` app works with "default" mode of `bthrm`. But I don't get readings for the `recorder` "Log BT HR" entry when using "both" or "default" mode of `bthrm`.

There should be no changes in how recording and events are handled. I have a look at this, did not explicitly test.

>     * If I run the `bthrm` app on the watch and then connect with the Web IDE I get errors and the reading seems to hang:
> 
> 
Hm, how do you get into debug mode? The code seems minified, maybe from message/messagegui judging by the `"txt"==` in there?

**Comment by thyttan** [comment]
> Hm, how do you get into debug mode?

I don't know. This is all after factory resetting the device and not doing anything other than installing and setting up the relevant apps - bthrm, recorder and run.

The debug thing seemed to happen only when bthrm were already running before connecting the web ide.

**Comment by halemmerich** [comment]
I can reproduce on 2v21, but only with fully updated apps. Some change outside of bthrm between the stock app versions and the current dev apploader behaves badly in combination with the new bthrm changes.

**Comment by halemmerich** [comment]
Wrapping the Bangle.emit in timeouts to decouple from the bthrm module scope helps with the debugger thing. The UI still stops working but `Bangle.on("BTHRM", print)` on the console works just fine on the console while the UI is frozen.

**Comment by halemmerich** [comment]
The last two commits seem to workaround the problem for me. It seems to happen during `prepareLazyRender` in the layout module, maybe there is some bug hiding there.

**Comment by thyttan** [comment]
Nice - I'll try it shortly!

We get this lint warn: 
```
 /home/runner/work/BangleApps/BangleApps/apps/bthrm/bthrm.js
Warning:   126:10  warning  'showStatusInfo' is defined but never used  no-unused-vars
```

But in the code:
```
// This can get called for the boot code to show what's happening
function showStatusInfo(txt) {
```
So it might be used somewhere else.

Maybe this should be added to the known warnings. Or if there's another way to permanently exempt just that warning, even if the app is updated?

**Comment by halemmerich** [comment]
Yes, that is used by the library code to write the tiny status info in the bottom of the BTHRM app. This will need an exemption. I'll try if the linter is happy if this method is put onto the global object explicitly.

**Comment by halemmerich** [comment]
> * The `recorder` app works with "default" mode of `bthrm`. But I don't get readings for the `recorder` "Log BT HR" entry when using "both" or "default" mode of `bthrm`.

This should be fixed now, there was an (old) typo preventing the recorder to start BTHRM on it's own.

**Comment by thyttan** [comment]
> I'll try if the linter is happy if this method is put onto the global object explicitly.

Seems to work! Would exporting the function have worked as well I wonder?

My watch does not reliably react to the hardware button when running the app now. I would suspect it's the changes to how `bthrm` uses the layout module that had made it so. Also the lock widget acts weird. And I'm a little unsure, but after a while the watch seems to get completely stuck until I hold down the hw button and kick the watchdog.

Now when connecting with the Web IDE while `bthrm` is running I can get this in the console:

```
Unable to retrieve board information. Connection Error?
Connected to Web Bluetooth, Bangle.js 92f9 (No response from board)
```

**Comment by halemmerich** [comment]
Exporting would probably have worked as well for the linter, but the boot code would then need to require the app which would not be great for RAM use if it is not in use.

The slowdown/hanging is very weird, I can see that sometimes too. That might come from the change to drawing every time an event arrives instead of once per second.

**Comment by halemmerich** [comment]
Layouting takes about 1 second, the events arrive at similar pace (but draw for every event) so the CPU is a 100% all the time. Lazy layout is even slower, that can't be right, can it?

**Comment by thyttan** [comment]
> Lazy layout is even slower, that can't be right, can it?

Certainly beats the purpose at least 🙃

**Comment by halemmerich** [comment]
I think I got it now. Updating the layout on every redraw is not needed, so the biggest chunk of CPU time was wasted anyway. This should behave a lot better. Funny how it was just on the border of working before with drawing once a second with ~1s draw times. It is now closer to 200ms per draw, much more reasonable.

**Comment by thyttan** [comment]
Everything seems to work now - and much better than before the refactor. Very good work! I think I should merge or do you want to do anything else before I do, @halemmerich ?

Just a small note (does not stop me merging the PR):
The `recorder` app works for getting HR readings from my Polar H10 now. It also add to headings `BT Battery,Energy expended,Contact,` - but I don't get readings for those. I don't know what this means, maybe other hr bt straps supplies them but not the H10.

**Comment by halemmerich** [comment]
I get Battery and Contact from mine. Did not yet see one that does Energy, but it is part of the same few bits delivered in the heart rate BLE characteristic so I just did implement it, it might well be incorrectly implemented but I think it is as the spec defines it. Since the headers and one additional comma per column are the only few bytes stored by the recorder I think it is not really a problem to store the empty columns. If it gets problematic they could be put behind a setting. Ideally including the code that handles the relevant characteristics in the library to save on processing as well.

This should be good for now, you can merge.

**Comment by thyttan** [comment]
Thank you for the great improvements!

**Comment by halemmerich** [comment]
Thanks for the detailed testing!

**Pull request merged**